### DISCUSSION TITLE
DynamoDB View repository example?

### DISCUSSION THREAD

**Discussion by CumpsD** [serverlesstechnology/cqrs#27] [Q&A]
The cqrs-demo repo has a postgress example. I was wondering if it was possible to add a dynamodb example too?

I currently am able to store commands in an event store in dynamodb, and was looking into using dynamodb lambda triggers to process new events for projections.

Should I create a View and then manually call `update` with a manual created `EventEnvelope`?

**Comment by CumpsD** [comment]
Trying out some idea's (not compiled, just in the GH editor) by gathering code from the examples:

```rust
use aws_sdk_dynamodb::Client;
use dynamo_es::DynamoViewRepository;

fn configure_view_repo(client: Client) -> DynamoViewRepository<MyView,MyAggregate> {
    DynamoViewRepository::new("my_view_table", client)
}

let config = aws_config::load_from_env().await;
let client = Client::new(&config);
let repo = Arc::new(configure_view_repo(client));
let query = GenericQuery::<MyViewRepository, MyView, MyAggregate>::new(repo);

... ?
```

I have been looking at https://github.com/serverlesstechnology/cqrs-demo/tree/master/src but it's not clear to me when `update` (https://github.com/serverlesstechnology/cqrs-demo/blob/master/src/queries.rs#L61) is called

**Reply by johnbcodes** [reply]
You might have moved on, but for others who come here later: `update` is only called by [`GenericQuery#apply_events`](https://github.com/serverlesstechnology/cqrs/blob/1d38c51624485affbf562850e4828ff668eb32d7/src/persist/generic_query.rs#L107) which is preceded by `GenericQuery#dispatch` which is called by `CqrsFramework#execute_with_metadata` in the `cqrs-demo` application. So I think your idea is correct and don't see how it would change with your rephrased question below if you have the events to apply.

**Reply by CumpsD** [reply]
Not moved on yet, just a lack of time :) I have a Lambda now which receives an SNS message over SQS with the events in it:

```rust
#[tokio::main]
async fn main() -> Result<(), Error> {
    run(service_fn(lambda_handler)).await?;

    Ok(())
}

async fn lambda_handler(
    event: LambdaEvent<SqsEventObj<SnsMessageObj<BankAccountEvent>>>,
) -> Result<(), Error> {
    println!("Context {:#?}", event.context);

    event
        .payload
        .records
        .into_iter()
        .for_each(|m| message_handler(m.body.message));

    Ok(())
}

fn message_handler(msg: BankAccountEvent) {
    match msg {
        BankAccountEvent::AccountOpened { .. } => println!("AccountOpened {:#?}", msg),
        BankAccountEvent::CustomerDepositedMoney { .. } => {
            println!("CustomerDepositedMoney {:#?}", msg)
        }
        BankAccountEvent::CustomerWithdrewCash { .. } => {
            println!("CustomerWithdrewCash {:#?}", msg)
        }
        BankAccountEvent::CustomerWroteCheck { .. } => println!("CustomerWroteCheck {:#?}", msg),
    };
}
```

But no parts of https://github.com/serverlesstechnology/cqrs in it yet

**Comment by CumpsD** [comment]
Perhaps I should rephrase my question to generalize it a bit more:

> Is it possible to use cqrs-es/dynamo-es from a stand-alone lambda to update a materialized view (projection in cqrs/es speak)?

That stand-alone lambda could then either be triggered from DynamoDB Streams, or Kinesis, or SNS triggering SQS triggering the lambda, ...

**Reply by davegarred** [reply]
My apologies, I missed this post originally.

As far as a dynamo demo, it should look very much like the setup in [cqrs-demo](https://github.com/serverlesstechnology/cqrs-demo) but swapping out individual components with their counterparts in [dynamo-ed](https://github.com/serverlesstechnology/dynamo-es).

I think you're on the right track, using SNS to distribute events to queries should work well. I have not tried this particular configuration but have used Kafka and Kinesis in a similar fashion.

I don't think I answered your question though?

**Reply by CumpsD** [reply]
I'll have to play with it, I managed to SNS part to distribute the events to standalone lambda's now. Now I need to figure out the parts needed from the cqrs-demo to update view projections, and try to get them to work together in that lambda, to accomplish this: https://doc.rust-cqrs.org/application_persisted_views.html
